docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,157 @@
+  * Each worker:
+    * Dequeues an item from *Dirty Queue*
+    * Checks if the item’s parents still exist. 
+    * Checks for a tombstone of parents that don’t exist
+    * If there is no parent nor tombstone, requests the API server to delete the item.
+    * If tombstone does exist, removes the parent from the object’s ParentReferences.
+
+* Propagator:
+  * The Propagator is for optimization, not for correctness.
+  * Maintains a DAG of parent-child relations. This DAG stores only name/uid/orphan triplets, not the entire body of every item.
+  * Consists of an *Event Queue* and a single worker.
+  * Watches for create/update/delete events for all resources, enqueues the events to the *Event Queue*.
+  * Worker:
+    * Dequeues an item from the *Event Queue*.
+    * If the item is an creation or update, then updates the DAG accordingly.
+      * If the object has a parent and the parent doesn’t exist in the DAG yet, then apart from adding the object to the DAG, also enqueues the object to the *Dirty Queue*.
+    * If the item is a deletion, then removes the object from the DAG, and enqueues all its children to the *Dirty Queue*.
+  * The propagator shouldn't need to do any RPCs, so a single worker should be sufficient. This makes locking easier.
+  * With the Propagator, we *only* need to run the Scanner when starting the Propagator to populate the DAG and the *Dirty Queue*.
+
+## Changes to existing components
+* Storage: we should add a REST storage for Tombstones. The index should be UID rather than namespace/name.
+
+* API Server: when handling a deletion request, if DeleteOptions.Orphaning is true, then the API Server either creates a tombstone with TTL if the tombstone doesn't exist yet, or updates the TTL of the existing tombstone. The API Server deletes the object after the tombstone is created.
+
+* Controllers: when creating child objects, controllers need to fill up the their ObjectMeta.ParentReferences field. Objects that don’t have a parent should have the namespace object as the parent.
+
+## Expected race conditions
+* *Controller of the parent resource creates a child object after GC has observed the deletion of the parent object*: There are two possible cases and both are well handled in this design. 
+  * If parent’s orphaning policy is false, the parent object should have already been removed from the DAG, so when the child object is created, it’s added to the *Dirty Queue* and will be processed by the Garbage Processor.
+  * If parent’s orphaning policy is true, a tombstone should exist in the DAG, so the newly created child object will be added as child of the tombstone, and no further action will be taken.
+    * Note that this requires the tombstone to exist for a long enough period.
+
+* *GC deletes the child object before noticing the updates to the ParentReferences*: If this race turns out to be a problem in practice, the deletion request should be sent with ResouceVersion as a precondition. 
+
```

Comment:
I think it would be good to add a section on how migration would work for existing data assuming this design?